### PR TITLE
Improve String#toUTF8

### PR DESCRIPTION
**Pull request by MaxGraey**
- Add optional argument for pre-calculated utf8 length.

### CODE REVIEW

File: std/assembly/string.ts [hunk 1]
```diff
@@ -432,8 +432,9 @@ export class String {
     return len;
   }
 
-  toUTF8(): usize {
-    var buf = memory.allocate(<usize>this.lengthUTF8);
+  toUTF8(len: i32 = -1): usize {
+    if (len == -1) len = this.lengthUTF8;
+    var buf = memory.allocate(<usize>len);
```
Comment (dcodeIO) [inline std/assembly/string.ts]: But, won't this write out of bounds below then?

Comment (MaxGraey) [reply std/assembly/string.ts]: Did you mean `MAX_LENGTH`?

Comment (dcodeIO) [reply std/assembly/string.ts]: No, in toUTF8, if you allocate let's say 16 bytes, but the string requires 32 bytes, with your change that'd write out of bounds.

Comment (MaxGraey) [reply std/assembly/string.ts]: Now I see. So wdyt about add additional check for loop condition?

Comment (MaxGraey) [reply std/assembly/string.ts]: Fixed. And add tests

**Comment by dcodeIO** [comment]
Thinking whether it makes sense to make a new function for variable length writing instead, like `String#writeUTF8(dest: usize, maxLength: usize)` that returns the actual number of bytes written. Wdyt?

**Comment by MaxGraey** [comment]
According to use cases which I saw `String#toUTF8` always used as pair with `String#lengthUT8` so for addressed and optimized this often case I created this PR. Not sure about `String#writeUTF8` because it duplicate `String#toUTF8` job with some flexibilities like `maxLength`. So if don't see this makes sense I better close this PR. Two methods can confused I prefer use minimalist api as possible

**Comment by dcodeIO** [comment]
I think there are two real world scenarios:

1. You want to convert a standard string **to** a new buffer (including the necessary allocation)
2. You want to **write** a standard string to an existing buffer (with possibly limited length)

In 1., an additional check for the anyway pre-allocated length is unnecessary overhead. Not sure if duplicating code there is much of a concern, because if the respective other option isn't used in a program, that function won't be compiled anyway. Hmm.

**Pull request closed without merging**